### DISCUSSION TITLE
snap to connect plc s7-1200 and read data

### DISCUSSION THREAD

**Discussion by gzy354293185** [gijzelaerr/python-snap7#341] [Q&A] [answered]
I want to use s7-1200plc to read data, this is my code

```
import snap7
plc = snap7.client.Client()
plc.connect('192.168.0.1', 0, 1)
data = plc.read_area(snap7.types.Areas.PA,0,0,1)
print(data)
```
and this is Error message
b'CPU : Function not available'
Traceback (most recent call last):
  File "E:\Holecount\plctest.py", line 21, in <module>
    plc.write_area(snap7.types.Areas.MK, 0, 2, a)
  File "E:\anaconda\envs\hole\lib\site-packages\snap7\client.py", line 26, in f
    check_error(code, context="client")
  File "E:\anaconda\envs\hole\lib\site-packages\snap7\common.py", line 92, in check_error
    raise Snap7Exception(error)
snap7.exceptions.Snap7Exception: b'CPU : Function not available'

**Comment by gzy354293185** [comment]
connect is success, just read data error

**Reply by swamper123** [reply]
Is your setup respecting [these 3 points](http://snap7.sourceforge.net/snap7_client.html#1200_1500)?:

```
1.    Only global DBs can be accessed.
2.    The optimized block access must be turned off.
3.    The access level must be “full” and the “connection mechanism” must allow GET/PUT.
```

**Reply by gzy354293185** [reply]
yes, I've revised everything

**Reply by swamper123** [reply]
Btw: Why is your code complaining about ` plc.write_area(snap7.types.Areas.MK, 0, 2, a)` in Line 26?
This is not an error in reading sth.!

**Reply by gzy354293185** [reply]
This is the error at line 26 in the  `client.py`  file

**Comment by swamper123** [comment]
Sry, line 21 in `File "E:\Holecount\plctest.py"`
This is what the traceback is talking about.
The lines below is just the function which checks, if an error was received.

**Reply by gzy354293185** [reply]
I just found that I uploaded the wrong code. I'll upload it again when I have time

**Comment by gijzelaerr** [comment, marked as answer]
Closing due to age and inactivity. For S7-1200 PLCs, ensure: (1) only global DBs are accessed, (2) optimized block access is turned off, (3) access level is 'full' with GET/PUT enabled.
